### DISCUSSION TITLE
How to execute Jobs if folder updated (recursive)?

### DISCUSSION THREAD

**Discussion by heathsnow** [community/community#25455] [Actions]
<p>Here’s my setup:</p>
<p><strong>folder structure</strong></p>
<pre><code class="lang-auto">repo:
  subdir:
    env1:
    env2:
</code></pre>
<p><strong>Workflow simple example I’d like to accomplish</strong></p>
<pre><code class="lang-auto">on:
  push:
    paths:
      - subdir/**
jobs:
  deploy:
    runs-on:  ubuntu-latest
    strategy:
      matrix:
        environment: [env1, env2]
    if: (execute only if ${{ matrix.env }} has changes)
</code></pre>
<p>I want a list of jobs to execute only for subdirectories that have commits.  Any help would be appreciated!</p>

**Comment by rabruce** [comment]
<p>IMO, your best path here is to have a previous job figuring out where the changes are, then setting the output and using that output to build your matrix job.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/heathsnow">@heathsnow</a>,</p>
<ol>
<li>If the commits pushed to the sub-directories occur before triggering the workflow, the paths filter is enough to meet your request. You just need to set some <a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#patterns-to-match-file-paths" rel="nofollow noopener">patterns</a> to match all the paths that you want to trigger the workflow runs on push event.</li>
</ol>
<pre><code class="lang-yaml">on:
  push:
    paths:
      - path-pattern-1
      - path-pattern-2
      . . .
      - path-pattern-N
</code></pre>
<ol start="2">
<li>If the commits to the sub-directories are pushed via a job in the workflow, you can set an output in this job to check the sub-directories are updated. If the the sub-directories are updated, set the value of the output to be <strong>true</strong>, otherwise <strong>false</strong>.<br>
A simple demo:</li>
</ol>
<pre><code class="lang-yaml">jobs:
  check_commits:
    runs-on: ubuntu-latest
    outputs:
      subdir_updated: ${{ steps.check_subdir.outputs.subdir_updated }}
    steps:
      - name: push commits
        . . .

      - name: check if sub-directories updated
        id: check_subdir
        run: |
        # here you need to replace "(sub-directories are updated)" with the actual if statement
          if [[ (sub-directories are updated) ]]; then
            echo "::set-output name=subdir_updated::true"
          else
            echo "::set-output name=subdir_updated::false"
          fi

  deploy:
    needs: [check_commits]
    if: needs.check_commits.outputs.subdir_updated == 'true'
    runs-on:  ubuntu-latest
    strategy:
      matrix:
        environment: [env1, env2]
    steps:
      . . .
</code></pre>

**Comment by heathsnow** [comment]
<aside class="quote group-github-partners" data-username="brightran" data-post="3" data-topic="117344">
<div class="title">
<div class="quote-controls"></div>
 brightran:</div>
<blockquote>
<p>If the commits pushed to the sub-directories occur before triggering the workflow, the paths filter is enough to meet your request.</p>
</blockquote>
</aside>
<p>I don’t think the paths filter is enough.  If I make a change to repo &gt; subdir &gt; env1 contents then I want a job that will only, for instance, have a step <code>echo "Updates made to env1"</code> but I don’t want to see that for env2 since no changes were made in that directory.</p>

**Comment by heathsnow** [comment]
<p>Not a bad idea, thanks.  I wish there were a canned solution already.  Really I’m just trying to work around the fact that GitHub doesn’t have templating yet.  I have 20+ workflows that are exactly the same except they trigger based on different paths getting updated and I was hoping to condense it down to 1 workflow with a matrix strategy.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/heathsnow">@heathsnow</a>,</p>
<p>You can use “git diff” command to list all the changed files, and then check if all the changed files are under the directory “<strong>subdir/env1</strong>”. If all the changed files are under that directory, set an output with the value is <strong>true</strong>, otherwise <strong>false</strong>. Then according to the value of the output to run or not run the subsequent job which has the step  <strong>echo “Updates made to env1”</strong>.<br>
A simple example:</p>
<pre><code class="lang-yaml">jobs:
  check:
    name: Check changed files
    outputs:
      run_job: ${{ steps.check_files.outputs.run_job }}
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v2
        with:
          fetch-depth: 2

      - name: check files
        id: check_files
        run: |
          echo "=============== list changed files ==============="
          git diff --name-only HEAD^ HEAD
          
          echo "========== check paths of changed files =========="
          git diff --name-only HEAD^ HEAD &gt; files.txt
          while IFS= read -r file
          do
            echo $file
            if [[ $file != "subdir/env1/"* ]]; then
              echo "This file is not under the directory 'subdir/env1'."
              echo "::set-output name=run_job::false"
              break
            else
              echo "::set-output name=run_job::true"
            fi
          done &lt; files.txt

  job_for_env1:
    name: Job for env1
    needs: check
    if: needs.check.outputs.run_job == 'true'
    runs-on: ubuntu-latest
    steps:
      - name: run for env1
        run: echo "Updates made to env1"
</code></pre>
